packages/internal/test/marshal.test.js

```diff
@@ -0,0 +1,301 @@
+      const val = marshaller.fromCapData(data);
+      valueHook?.(val, 'from');
+      return val;
+    },
+  });
+};
+
+/**
+ * @param {import('ava').ExecutionContext} t
+ * @param {any} val
+ */
+const checkRemoteMarshallerValueInvariants = (t, val) => {
+  t.true(Array.isArray(val));
+  t.true(val.length > 0);
+  let hasCap = false;
+  for (const elem of val) {
+    const passStyle = passStyleOf(elem);
+    switch (passStyle) {
+      case 'null':
+      case 'undefined':
+        break;
+      case 'remotable':
+        hasCap = true;
+        break;
+      default:
+        t.fail(`Unexpected pass-style: ${passStyle}`);
+    }
+  }
+  t.true(hasCap);
+};
+
+const unexpectedMarshallerInvocation = (t, val) => {
+  t.log('Unexpected marshalled value', val);
+  t.fail('Remote marshaller should not have been invoked');
+};
+
+test('wrapRemoteMarshaller - empty slots', async t => {
+  const marshaller = makeMockMarshaller({
+    valueHook: val => {
+      unexpectedMarshallerInvocation(t, val);
+    },
+  });
+
+  const wrappedMarshaller = wrapRemoteMarshaller(marshaller);
+
+  const specimen = harden({ foo: 42, err: Error('bar') });
+
+  const capData = await wrappedMarshaller.toCapData(specimen);
+  const clone = await wrappedMarshaller.fromCapData(capData);
+
+  t.deepEqual(clone, specimen);
+});
+
+const withSlots = test.macro(async (t, { withCache } = {}) => {
+  let valueHookCalled = false;
+  const marshaller = makeMockMarshaller({
+    valueHook: val => {
+      if (valueHookCalled && withCache) {
+        unexpectedMarshallerInvocation(t, val);
+      } else {
+        valueHookCalled = true;
+        checkRemoteMarshallerValueInvariants(t, val);
+      }
+    },
+  });
+  const wrappedMarshaller = wrapRemoteMarshaller(
+    marshaller,
+    {},
+    withCache
+      ? {} // default caches
+      : {
+          slotToVal: null,
+          valToSlot: null,
+        },
+  );
+
+  const specimen = harden({ foo: 42, bar: Far('bar', { sentinel() {} }) });
+
+  // Will populate the cache
+  const capData = await wrappedMarshaller.toCapData(specimen);
+  // Will hit the cache
+  const clone = await wrappedMarshaller.fromCapData(capData);
+
+  t.deepEqual(clone, specimen);
+
+  // Will hit the cache
+  const capData2 = await wrappedMarshaller.toCapData(specimen);
+
+  const caches = {
+    slotToVal: withCache ? new Map() : null,
+    valToSlot: withCache ? new Map() : null,
+  };
+  const wrappedMarshaller2 = wrapRemoteMarshaller(marshaller, {}, caches);
+  valueHookCalled = false;
+
+  // Populate the cache
+  const clone2 = await wrappedMarshaller2.fromCapData(capData2);
+  t.true(valueHookCalled);
+  t.deepEqual(clone2, specimen);
+
+  if (withCache) {
+    const cap = specimen.bar;
+    const slot = capData2.slots[0];
+    t.is(caches.slotToVal?.get(slot), cap);
+    t.is(caches.valToSlot?.get(cap), slot);
+  }
+
+  // Hit the cache
+  const clone3 = await wrappedMarshaller2.fromCapData(capData2);
+  t.deepEqual(clone3, specimen);
```

Comment:
`t.deepEqual` isn't sufficient to establish that remotable identity (`Far('bar'...)`) is preserved

Reply:
That actually isn't accurate. `deepEqual` will check that 2 functions have the same identity, which is what the `sentinel` in the `Far('bar', { sentinel() {} })` provides. So while technically the remotable could be a new one with the same sentinel function, we rely on the fact that no code actually mangles with remotables themselves.